### ISSUE TITLE
[BUG] ranking metrics should either accept string users, item ids or complain about it

### ISSUE THREAD

**Issue by miguelgfierro** [Microsoft/Recommenders#543]
### Description

Given this scenario:

```
print(top_k_scores.dtypes)
UserId         object
MovieId        object
prediction    float64
dtype: object

print(test.dtypes)
UserId         int64
MovieId        int64
Rating       float64
Timestamp      int64
dtype: object

map_at_k(test, top_k_scores, k=10)
ndcg_at_k(test, top_k_scores, k=10)
...

```
we will get 0 in all of them. If we transform the top_k_scores string columns to numeric, it works. 

### In which platform does it happen?

Azure Data Science Virtual Machine

### Expected behavior (i.e. solution)

Preferably, if the columns doesn't have the correct type, it should raise an exception. Alternatively, we can convert everything to numeric.

### Other Comments
@yueguoguo what are your thoughts on this? Is it easy to combine str and numeric in the ranking computation?

**Comment by miguelgfierro** [comment]
This happened to me when working with SAR. In this line https://github.com/Microsoft/Recommenders/blob/master/reco_utils/recommender/sar/sar_singlenode.py#L310 we are converting the item and user columns to str. In fact, in the notebook, we are converting back to numeric

**Comment by gramhagen** [comment]
yeah it would be great to remove the conversion of types in sar and rely on doing that if necessary in the metrics calculations. I believe this happens in the spark evaluation as well, most likely due to a join between the prediction and truth dataframes that fails if the types are different.

**Comment by yueguoguo** [comment]
Just noticed @gramhagen 's comment. I would prefer raising errors so that the users can handle dtype coercions by themselves. Without any prior knowledge about the data, it might be dangerous to perform the type conversion because it might lead to incorrect calculation of the metrics. Our Spark implementation now practices the type conversion, which is inconsistent here with the Python implementation. After we reach an agreement, I can change both to make sure they are consistent. 

What do you guys think? 
+ @anargyri @loomlike

**Comment by yueguoguo** [comment]
Also noticed that the PR did not pass because the raised exception is captured in running some notebook testing, i.e., surprise and vw.

**Comment by anargyri** [comment]
My comment is that all the algorithms and methods that accept userID and/or itemID should be able to accept these as strings. Because this is the typical case in the logs or DB of a company, they are encoded as alphanumeric strings.  So, I think the most user-friendly approach is to allow the user to input strings in the dataframe, as we currently do.
I think the line in the SAR algorithm that Miguel mentioned addresses the case (which can also occur) that the user inputs a dataframe with integer user or item IDs. But there is probably some redundancy in the notebook or in the SAR methods.

**Comment by yueguoguo** [comment]
> My comment is that all the algorithms and methods that accept userID and/or itemID should be able to accept these as strings. Because this is the typical case in the logs or DB of a company, they are encoded as alphanumeric strings. So, I think the most user-friendly approach is to allow the user to input strings in the dataframe, as we currently do.

Yea we allow string typed input columns, but not columns with mismatched types, e.g., string in true and numeric in prediction. 

> I think the line in the SAR algorithm that Miguel mentioned addresses the case (which can also occur) that the user inputs a dataframe with integer user or item IDs. But there is probably some redundancy in the notebook or in the SAR methods.

Right. Now the discussion is about where we put the "redundancy", algorithm, evaluator, or the notebook (this means user handles it by themselves). 😄

**Comment by miguelgfierro** [comment]
I think whatever we choose (input as str or int) we should make it explicit, so the user knows that he or she has to use an specific type

**Comment by yueguoguo** [comment]
Resolved in #553
